src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1442,16 +1442,45 @@ public Solution WithAnalyzerConfigDocumentTextLoader(DocumentId documentId, Text
         }
 
         /// <summary>
-        /// Creates a branch of the solution that has its compilations frozen in whatever state they are in at the time, assuming a background compiler is
-        /// busy building this compilations.
-        /// 
-        /// A compilation for the project containing the specified document id will be guaranteed to exist with at least the syntax tree for the document.
-        /// 
-        /// This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
+        /// Creates a branch of the solution that has its compilations frozen in whatever state they are in at the time,
+        /// assuming a background compiler is busy building this compilations.
+        /// <para/> A compilation for the project containing the specified document id will be guaranteed to exist with
+        /// at least the syntax tree for the document.
+        /// <para/> This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
         /// </summary>
         internal Solution WithFrozenPartialCompilationIncludingSpecificDocument(DocumentId documentId, CancellationToken cancellationToken)
         {
-            var newCompilationState = _compilationState.WithFrozenPartialCompilationIncludingSpecificDocument(documentId, cancellationToken);
+            // in progress solutions are disabled for some testing
+            if (this.Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
+                return this;
+
+            var currentCompilationState = _compilationState;
+            var currentDocumentState = this.SolutionState.GetRequiredDocumentState(documentId);
+
+            // We want all linked versions of this document to also be present in the frozen solution snapshot (that way
+            // features like 'completion' can see that there are linked docs and give messages about symbols not being
+            // available in certain project contexts). We do this in a slightly hacky way for perf though. Specifically,
+            // instead of parsing *all* the sibling files (which can be expensive, especially for a file linked in many
+            // projects/tfms), we only parse this single tree.  We then use that same tree across all siblings.  That's
+            // technically inaccurate, but we can accept that as the primary purpose of 'frozen partial' is to get a
+            // snapshot *fast* that is allowed to be *inaccurate*.
```

Comment:
So what worries me here with this example for completion is it'll potentially break it, right? Consider if you have a multi-targeting project doing something like:

```
#if NET
// I can type here and get completion for the .NET Core target
#else
// I can type here and get completion for the .NET Framework target
#endif
```

Right now you can type in either side and because we look at linked files we'll still have correct completion, even though I don't have to use the switcher. Now that won't work anymore, since if you're typing in the non-active area, you'll be in directive trivia now instead of the regular stuff.

How esoteric is this case? I don't know. And maybe this is something we just get in early enough and wait for feedback.

I can imagine one potential way to make this a bit smarter for completion, at least: if the frozen partial helpers knew about the position that's going to later be interrogated, they could be a bit smarter by saying if that location is within a #if then maybe we can't share trees, but if it's outside, then share like you're implementing. Or put another way, the "force" flag becomes a bit smarter to say force if you're not inside a directive.

Reply:
So the above should work (and i'll test for it).  That's because you always get the correct tree for the *current* context.  It's just that the sibling trees will not be correct for their particular contexts.

But htat's *ok*.  We don't actually ever do anything *within* those trees.  In other words, completion really only uses those other contexts to determine which symbols (like if a method off of a type is available in netfx vs netcore) are available or not.  That's really not something that will realistically be impacted by the trees.